samples/subsys/usb/mass/src/main.c

```diff
@@ -1,119 +1,155 @@
 		printk("Erasing flash area ... ");
 		rc = flash_area_erase(pfa, 0, pfa->fa_size);
 		printk("%d\n", rc);
 	}
 
-	rc = fs_mount(&fs_mnt);
+	if (rc != 0) {
+		flash_area_close(pfa);
+	}
+#endif
+	return rc;
+}
+
+static int mount_app_fs(struct fs_mount_t *mnt)
+{
+	int rc;
+
+#if CONFIG_FAT_FILESYSTEM_ELM
+	static FATFS fat_fs;
+
+	mnt->type = FS_FATFS;
+	mnt->fs_data = &fat_fs;
+	if (IS_ENABLED(CONFIG_DISK_ACCESS_RAM)) {
+		mnt->mnt_point = "/RAM:";
+	} else {
+		mnt->mnt_point = "/NAND:";
+	}
+
+#elif CONFIG_FILE_SYSTEM_LITTLEFS
+	mnt->type = FS_LITTLEFS;
+	mnt->mnt_point = "/lfs";
+	mnt->fs_data = &storage;
+#endif
+	rc = fs_mount(mnt);
+
+	return rc;
+}
+
+static void setup_disk(void)
+{
+	struct fs_mount_t *mp = &fs_mnt;
+	int rc;
+
+	if (IS_ENABLED(CONFIG_DISK_ACCESS_FLASH)) {
+		rc = setup_flash(mp);
+		if (rc != 0) {
+			LOG_ERR("Failed to setup flash area");
+			return;
+		}
+	}
+
+	if (!IS_ENABLED(CONFIG_FILE_SYSTEM_LITTLEFS) &&
+	    !IS_ENABLED(CONFIG_FAT_FILESYSTEM_ELM)) {
+		LOG_INF("No file system selected");
+		return;
+	}
+
+	rc = mount_app_fs(mp);
+	if (rc != 0) {
+		LOG_ERR("Failed to mount filesystem");
+		return;
+	}
 
 	/* Allow log messages to flush to avoid interleaved output */
 	k_sleep(K_MSEC(50));
 
 	printk("Mount %s: %d\n", fs_mnt.mnt_point, rc);
 
-	if (rc >= 0) {
-		struct fs_statvfs sbuf;
+	struct fs_statvfs sbuf;
+
+	rc = fs_statvfs(mp->mnt_point, &sbuf);
+	if (rc < 0) {
+		printk("FAIL: statvfs: %d\n", rc);
+		return;
+	}
+
+	printk("%s: bsize = %lu ; frsize = %lu ;"
+	       " blocks = %lu ; bfree = %lu\n",
+	       mp->mnt_point,
+	       sbuf.f_bsize, sbuf.f_frsize,
+	       sbuf.f_blocks, sbuf.f_bfree);
+
+	struct fs_dir_t dir = { 0 };
 
-		rc = fs_statvfs(mp->mnt_point, &sbuf);
+	rc = fs_opendir(&dir, mp->mnt_point);
+	printk("%s opendir: %d\n", mp->mnt_point, rc);
+
+	while (rc >= 0) {
+		struct fs_dirent ent = { 0 };
+
+		rc = fs_readdir(&dir, &ent);
 		if (rc < 0) {
-			printk("FAIL: statvfs: %d\n", rc);
-			goto out;
+			break;
 		}
-
-		printk("%s: bsize = %lu ; frsize = %lu ;"
-		       " blocks = %lu ; bfree = %lu\n",
-		       mp->mnt_point,
-		       sbuf.f_bsize, sbuf.f_frsize,
-		       sbuf.f_blocks, sbuf.f_bfree);
-
-		struct fs_dir_t dir = { 0 };
-
-		rc = fs_opendir(&dir, mp->mnt_point);
-		printk("%s opendir: %d\n", mp->mnt_point, rc);
-
-		while (rc >= 0) {
-			struct fs_dirent ent = { 0 };
-
-			rc = fs_readdir(&dir, &ent);
-			if (rc < 0) {
-				break;
-			}
-			if (ent.name[0] == 0) {
-				printk("End of files\n");
-				break;
-			}
-			printk("  %c %u %s\n",
-			       (ent.type == FS_DIR_ENTRY_FILE) ? 'F' : 'D',
-			       ent.size,
-			       ent.name);
+		if (ent.name[0] == 0) {
+			printk("End of files\n");
+			break;
 		}
-
-		(void)fs_closedir(&dir);
+		printk("  %c %u %s\n",
+		       (ent.type == FS_DIR_ENTRY_FILE) ? 'F' : 'D',
+		       ent.size,
+		       ent.name);
 	}
 
-out:
-	flash_area_close(pfa);
-#endif
+	(void)fs_closedir(&dir);
```

Comment:
Maybe printk some error if we get here with `rc < 0`; note that the we can get `rc < 0` from inside of the `while` loop at line 133 and `fs_opendir` at line 130.

Reply:
@jfischer-no please address